### PR TITLE
Alpha Proposal

### PR DESCRIPTION
**Pull request by ElliotFriedman**
Create proposal structure as defined in issue [45](https://github.com/solidity-labs-io/forge-proposal-simulator/issues/45).

TODO:

- [x] update docs

### CODE REVIEW

File: examples/timelock/TIMELOCK_01.sol [hunk 1]
```diff
@@ -19,6 +21,16 @@ contract TIMELOCK_01 is TimelockProposal {
...
     }
 
+    /// @notice always reverts, do not use this method for timelock proposals
+    function getCalldata()
+        public
+        view
+        override(Proposal, TimelockProposal)
+        returns (bytes memory data)
+    {
+        return TimelockProposal.getCalldata();
```
Comment (anajuliabit) [inline examples/timelock/TIMELOCK_01.sol:31]: what is the point of overriding here?

File: proposals/AlphaProposal.sol [hunk 2]
```diff
@@ -0,0 +1,100 @@
+pragma solidity ^0.8.0;
+
+import {Strings} from "@openzeppelin/utils/Strings.sol";
+
+import {VmSafe} from "@forge-std/Vm.sol";
+
+import {Proposal} from "@proposals/Proposal.sol";
+import {Addresses} from "@addresses/Addresses.sol";
+
+abstract contract AlphaProposal is Proposal {
```
Comment (anajuliabit) [inline proposals/AlphaProposal.sol:10]: wouldn't it be better to implement this directly in the Proposal class for a cleaner solution?

File: proposals/AlphaProposal.sol [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        bytes memory data
+    ) private pure returns (string memory) {
+        /// Initialize an array of characters twice the length of data,
+        /// since each byte will be represented by two hexadecimal characters
+        bytes memory buffer = new bytes(data.length * 2);
+
+        /// Characters for conversion
+        bytes memory characters = "0123456789abcdef";
+
+        for (uint i = 0; i < data.length; i++) {
```
Comment (anajuliabit) [inline proposals/AlphaProposal.sol:91]:
```suggestion
        for (uint256 i = 0; i < data.length; i++) {
```

File: proposals/AlphaProposal.sol [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            /// For each byte, find the corresponding hexadecimal characters
+            buffer[i * 2] = characters[uint(uint8(data[i] >> 4))];
```
Comment (anajuliabit) [inline proposals/AlphaProposal.sol:93]:
```suggestion
            buffer[i * 2] = characters[uint256(uint8(data[i] >> 4))];
```

File: proposals/AlphaProposal.sol [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            buffer[i * 2 + 1] = characters[uint(uint8(data[i] & 0x0f))];
```
Comment (anajuliabit) [inline proposals/AlphaProposal.sol:94]:
```suggestion
            buffer[i * 2 + 1] = characters[uint256(uint8(data[i] & 0x0f))];
```

File: proposals/TimelockProposal.sol [hunk 6]
```diff
@@ -8,6 +8,10 @@ import {Address} from "@utils/Address.sol";
 abstract contract TimelockProposal is Proposal {
     using Address for address;
 
+    function getCalldata() public view virtual override returns (bytes memory) {
```
Comment (anajuliabit) [inline proposals/TimelockProposal.sol:11]: natspec

File: proposals/TimelockProposal.sol [hunk 7]
```diff
@@ -8,6 +8,10 @@ import {Address} from "@utils/Address.sol";
...
+        revert("do not call getcalldata on timelock proposal");
```
Comment (anajuliabit) [inline proposals/TimelockProposal.sol:12]: why can't it be called?

**Review by anajuliabit — CHANGES_REQUESTED** [review]
I loved this, LFG! Just minor things:

File: examples/multisig/MULTISIG_01.sol [hunk 8]
```diff
@@ -1,14 +1,16 @@
...
 import {MockToken} from "@examples/MockToken.sol";
+import {Addresses} from "@addresses/Addresses.sol";
+import {AlphaProposal} from "@proposals/AlphaProposal.sol";
+import {MultisigProposal} from "@proposals/MultisigProposal.sol";
 
 // MULTISIG_01 proposal deploys a Vault contract and an ERC20 token contract
 // Then the proposal transfers ownership of both Vault and ERC20 to the multisig address
 // Finally the proposal whitelist the ERC20 token in the Vault contract
-contract MULTISIG_01 is MultisigProposal {
+contract MULTISIG_01 is AlphaProposal, MultisigProposal {
```
Comment (prateek105) [inline examples/multisig/MULTISIG_01.sol:13]: `buildModifier` not used for this proposal. If we are keeping this as a simple proposal then we can avoid inheriting `AlphaProposal` contract here.

File: examples/timelock/TIMELOCK_01.sol (hunk 1 above)
Comment (anajuliabit) [reply examples/timelock/TIMELOCK_01.sol:31]: would it be better to implement AlphaProposal in the Proposal so we don't need to have 2 parent classes?

File: examples/governor-bravo/BRAVO_02.sol [hunk 9]
```diff
@@ -1,45 +1,43 @@
 pragma solidity ^0.8.0;
 
-import {GovernorBravoProposal} from "@proposals/GovernorBravoProposal.sol";
-import {Addresses} from "@addresses/Addresses.sol";
 import {Vault} from "@examples/Vault.sol";
+import {Proposal} from "@proposals/Proposal.sol";
```
Comment (anajuliabit) [inline examples/governor-bravo/BRAVO_02.sol:4]: remove

File: examples/governor-bravo/BRAVO_03.sol [hunk 10]
```diff
@@ -1,56 +1,61 @@
 pragma solidity ^0.8.0;
 
-import {GovernorBravoProposal} from "@proposals/GovernorBravoProposal.sol";
-import {Addresses} from "@addresses/Addresses.sol";
 import {Vault} from "@examples/Vault.sol";
+import {Proposal} from "@proposals/Proposal.sol";
```
Comment (anajuliabit) [inline examples/governor-bravo/BRAVO_03.sol:4]: remove

File: examples/governor-bravo/BRAVO_04.sol [hunk 11]
```diff
@@ -0,0 +1,68 @@
+pragma solidity ^0.8.0;
+
+import {Vault} from "@examples/Vault.sol";
+import {MockToken} from "@examples/MockToken.sol";
+import {Addresses} from "@addresses/Addresses.sol";
+import {GovernorBravoProposal} from "@proposals/GovernorBravoProposal.sol";
+
+/// @notice Mock proposal that withdraws MockToken from Vault
+/// using the AlphaProposal type.
```
Comment (anajuliabit) [inline examples/governor-bravo/BRAVO_04.sol:9]:
```suggestion
/// using the GovernorBravoProposal type.
```

File: examples/multisig/MULTISIG_02.sol [hunk 12]
```diff
@@ -1,55 +1,53 @@
 pragma solidity ^0.8.0;
 
-import {MultisigProposal} from "@proposals/MultisigProposal.sol";
-import {Addresses} from "@addresses/Addresses.sol";
 import {Vault} from "@examples/Vault.sol";
+import {Proposal} from "@proposals/Proposal.sol";
```
Comment (anajuliabit) [inline examples/multisig/MULTISIG_02.sol:4]: remove

File: examples/multisig/MULTISIG_03.sol [hunk 13]
```diff
@@ -1,38 +1,42 @@
 pragma solidity ^0.8.0;
 
-import {MultisigProposal} from "@proposals/MultisigProposal.sol";
-import {Addresses} from "@addresses/Addresses.sol";
 import {Vault} from "@examples/Vault.sol";
+import {Proposal} from "@proposals/Proposal.sol";
```
Comment (anajuliabit) [inline examples/multisig/MULTISIG_03.sol:4]: remove

**Review by anajuliabit — CHANGES_REQUESTED** [review]

**Review by anajuliabit — APPROVED** [review]
LGTM

File: proposals/AlphaProposal.sol [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+import {Addresses} from "@addresses/Addresses.sol";
+
+abstract contract AlphaProposal is Proposal {
+    using Strings for *;
+
+    /// @notice starting snapshot of the contract state before the calls are made
+    uint256 private _startSnapshot;
+
+    /// @notice to be used by the build function easily create a governance proposal
+    modifier buildModifier(address caller, Addresses addresses) {
```
[the remark this answers is not in the document]
Comment (prateek105) [reply proposals/AlphaProposal.sol]: yes right, we can call the `_outerBuild` in `run`

**Review by prateek105 — APPROVED** [review]
lgtm.

**Pull request merged**